akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala

```diff
@@ -466,6 +466,94 @@ final class Broadcast[T](private val outputPorts: Int, eagerCancel: Boolean) ext
 
 }
 
+object Partition {
+
+  /**
+   * Create a new `Partition` stage with the specified input type.
+   *
+   * @param outputPorts number of output ports
+   * @param partitioner function deciding which output each element will be targeted
+   */
+  def apply[T](outputPorts: Int, partitioner: T ⇒ Int): Partition[T] = new Partition(outputPorts, partitioner)
+}
+
+/**
+ * Fan-out the stream to several streams. emitting an incoming upstream element to one downstream consumer according
+ * to the partitioner function applied to the element
+ *
+ * '''Emits when''' emits when an element is available from the input and the chosen output has demand
+ *
+ * '''Backpressures when''' the currently chosen output back-pressures
+ *
+ * '''Completes when''' upstream completes and no output is pending
+ *
+ * '''Cancels when'''
+ *   when all downstreams cancel
+ */
+
+final class Partition[T](outputPorts: Int, partitioner: T ⇒ Int) extends GraphStage[UniformFanOutShape[T, T]] {
+
+  val in: Inlet[T] = Inlet[T]("Partition.in")
+  val out: Seq[Outlet[T]] = Seq.tabulate(outputPorts)(i ⇒ Outlet[T]("Partition.out" + i))
+  override val shape: UniformFanOutShape[T, T] = UniformFanOutShape[T, T](in, out: _*)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    private var outPending: Option[(Int, T)] = None
+
+    setHandler(in, new InHandler {
+      override def onPush() = {
+        val elem = grab(in)
+        val idx = partitioner(elem)
+        if (isAvailable(out(idx))) {
+          if (!isClosed(out(idx))) {
+            push(out(idx), elem)
+            if (out.exists(x ⇒ isAvailable(x)) && !hasBeenPulled(in))
+              pull(in)
+          }
+        } else {
+          outPending = Some(idx -> elem)
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        if (!outPending.isDefined)
+          completeStage()
+      }
+    })
+
+    out.zipWithIndex.foreach {
+      case (o, idx) ⇒
+        setHandler(o, new OutHandler {
+          override def onPull() = {
+            outPending match {
+              case Some((i, elem)) if (i == idx) ⇒
+                push(o, elem)
+                outPending = None
+                if (!hasBeenPulled(in)) {
+                  pull(in)
+                }
+              case None ⇒
+                if (outPending.isEmpty && !hasBeenPulled(in)) {
+                  pull(in)
+                }
+            }
+          }
+
+          override def onDownstreamFinish(): Unit = {
+            if (out.forall(isClosed(_)))
```

Comment:
Maybe I am missing something, but, besides terminating when all outputs are closed, since I have a outPending counter, I have to update it when an OUT is closing and since I don't have a Boolean per OUT anymore, but rather a global counter, when an OUT is leaving I don't know whether it was pending or not since at this stage I cannot use isAvailable(out) ... hence I have to go back to a flag per out instead of the global pending counter.